### PR TITLE
Guych/v3 infra

### PR DESCRIPTION
**Pull request by guykanyvision**

### CODE REVIEW

**Review by orihabAnyvision — DISMISSED** [review]
nothing blocking here so - approved.

File: automation/devops_automation_infra/k8s_plugins/proxy_daemonset.py [hunk 1]
```diff
@@ -57,6 +60,7 @@ def _k8s_v1_client(self):
     def run(self):
         self.kill()
         logging.debug("Deploying automation-proxy DaemonSet")
+        kubectl.create_image_pull_secret(self._cluster.Kubectl.client())
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/k8s_plugins/proxy_daemonset.py:47]: this is idempotent right ?

Comment (guykanyvision) [reply automation/devops_automation_infra/k8s_plugins/proxy_daemonset.py:47]: yes

File: automation/devops_automation_infra/k8s_plugins/proxy_daemonset.py [hunk 2]
```diff
@@ -8,7 +9,9 @@
 from devops_automation_infra.plugins.tunnel_manager import TunnelManager
 from infra.model import cluster_plugins
 import kubernetes
+from devops_automation_infra.k8s_plugins.kubectl import Kubectl
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/k8s_plugins/proxy_daemonset.py:11]: this is reduntant i think

File: automation/devops_automation_infra/utils/kubectl.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    sec = kubernetes.client.V1Secret()
+    sec.metadata = kubernetes.client.V1ObjectMeta(name=name)
+    sec.type = type
+    sec.data = data
+    v1.create_namespaced_secret(namespace=namepace, body=sec)
+
+
+def create_image_pull_secret(client, config_path=None):
+    if not config_path:
+        config_path = f"{os.environ['userhome']}/.docker/config.json"
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/utils/kubectl.py:26]: NIT: 
``` python
config_path = config_path or f"{os.environ['userhome']}/.docker/config.json"
```

File: automation/devops_automation_infra/utils/kubectl.py [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    with open(config_path, 'rb') as f:
+        data = {'.dockerconfigjson': base64.b64encode(f.read()).decode()}
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/utils/kubectl.py:28]: NIT:
can you please rename it to docker_images_creds .. otherwise it is not clear what is this in this file and why do we add it as secret

File: automation/devops_automation_infra/utils/kubectl.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    try:
+        create_generic_secret(client=client, namepace='default', name='imagepullsecret', type='kubernetes.io/dockerconfigjson', data=data)
+    except ApiException as e:
+        if e.status == 409:
+            logging.debug("imagepullsecret is already exists")
+        else:
+            raise e
+
+
+def is_stateful_set_ready(client, name, namespace='default'):
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/utils/kubectl.py:39]: why is this related to the commit ?

File: automation/devops_automation_infra/k8s_plugins/kafka.py [hunk 6]
```diff
@@ -2,25 +2,48 @@
...
         self._name = 'kafka-cluster'
-        self._is_exposed = False
+
+    @property
+    def _is_running(self):
+        return kubectl.is_stateful_set_ready(name='kafka-cluster-kafka', client=self._cluster.Kubectl.client())
+
+    def _list_broker_pods(self):
+        return kubectl.get_pods_by_label(self._cluster.Kubectl.client(),
+                                         namespace=self._namespace, label='strimzi.io/name=kafka-cluster-kafka')
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/k8s_plugins/kafka.py:25]: NIT: 
i would extract the LABEL to class variable .. like K8S_SERVICE_LABEL

File: automation/devops_automation_infra/k8s_plugins/kafka.py [hunk 7]
```diff
@@ -30,7 +53,18 @@ def _expose(self):
...
+        pods_timestamps = [pod.status.start_time for pod in self._list_broker_pods()]
+        custom_object_client.patch_namespaced_custom_object(namespace=self._namespace,
+                                                            group='kafka.strimzi.io',
+                                                            version='v1beta1',
+                                                            plural='kafkas',
+                                                            name=self._name,
+                                                            body={'spec': kafka_spec})
+
+        logging.debug("Waiting for kafka brokers to restart")
+        waiter.wait_for_predicate(lambda: [pod.status.start_time for pod in self._list_broker_pods()] != pods_timestamps, timeout=30)
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/k8s_plugins/kafka.py:66]: there is an issue with this method . what happens if one broker restatred 
i think this should be extracted to 
```
brokers_state = self._brokers_state()
HERE the restart code .. 

``` python
waiter.wait_for_predicate(lambda: _kafka_brokers_restarted(brokers_state)

def _brokers_state():
    {pod.metadata.name : pod.status.start_time}

def _kafka_brokers_restarted(old_brokers_state):
    state=self._brokers_state()
    return len(old_brokers_state) == len(state) and len(a1.items() & b1.items()) == 0
```

**Review by sashas-anyvision — COMMENTED** [review]
look at the comments but in general very good

**Review by sashas-anyvision — APPROVED** [review]

File: automation/devops_automation_infra/tests/test_kafka.py [hunk 8]
```diff
@@ -10,12 +13,25 @@ def example_decorator():
     return True
 
 
+# @hardware_config(hardware={"host1": {"hardware_type": "ai_camera"}},
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/tests/test_kafka.py:16]: lets remove this

File: automation/devops_automation_infra/utils/kafka.py [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+    # Lazy initiate of topic assignment
+    consumer.topics()
+    waiter.wait_for_predicate(lambda: consumer.beginning_offsets(consumer.assignment()) ==
+                                      consumer.end_offsets(consumer.assignment()))
+
+    update_topic_config(admin, name, {"retention.ms": retention})
+    consumer.close()
+
+
+def get_topic_config_value(admin, topic_name, config_key):
```
Comment (sashas-anyvision) [inline automation/devops_automation_infra/utils/kafka.py:27]: consider return the entries

**Review by sashas-anyvision — APPROVED** [review]
